README.md
# ML_learning

## Run command : streamlit run app.py
## Create a .env file with OPENAI_API_KEY in it.
## Chat History comes from history.txt file, if you want to reset the history , for now manually delete the data from this file.

### install requirments as given in requirments.txt file.


history.txt
[{"You:": "what are industry challenges", "LLM:": "The industry challenges mentioned in the given context include cloud cost, data security, privacy, governance, AI disruption in existing businesses, and the modern way to support and scale data."}, {"You:": "what is marlin", "LLM:": "Marlin is a track or component within the Data and Analytics devops process. It focuses on defining the ingestion framework and building foundational components for data ingestion from various sources and systems. Marlin's capabilities include flatfile ingestion, Sqoop ingestion, JDBC ingestion, Salesforce ingestion, RDBMS ingestion, MongoDB ingestion, and more."}, {"You:": "what is marlin", "LLM:": "Marlin is a track or component within the Data and Analytics devops process. It focuses on defining the ingestion framework and building foundational components for data ingestion from various sources and systems. Marlin's capabilities include flatfile ingestion, Sqoop ingestion, JDBC ingestion, Salesforce ingestion, RDBMS ingestion, MongoDB ingestion, and more."}, {"You:": "what are the revenue projections for 2025", "LLM:": "The revenue projections for 2025 are not provided in the given context."}, {"You:": "what are revenue projection", "LLM:": "The revenue projection for the company, Decision Minds, is $28 million for 2023, $32 million for 2024, and $40 million for 2025."}, {"You:": "what are revenue projection for 2025", "LLM:": "The revenue projection for 2025 is $40."}, {"You:": "what is marlin", "LLM:": "The text does not provide information on what \"marlin\" is."}, {"You:": "what is marlin", "LLM:": "Marlin is a track or project that focuses on defining the Ingestion Framework and designing and building foundational components and modular, reusable utilities for data ingestion from various types of sources and systems. It includes a Common Vault framework to get secrets/passwords from Vault in a secured way and a Common DBConnection framework that currently supports PRESTO, HANA, HIVE, HAWQ databases. Marlin's capabilities also include various types of data ingestion such as Flatfile, Sqoop, JDBC, Salesforce, RDBMS, MongoDB, and more."}, {"You:": "what is Presto", "LLM:": "Presto is a Distributed SQL Query Engine. It is one of the tools implemented as a part of the Tesseract Platform and is accessible to Data Engineers."}, {"You:": "summarize previous answer.", "LLM:": "The text discusses various aspects of VMware's Tesseract platform. Iceberg is the main portal for data needs and research, allowing users to search for data sets, integrate with data catalogs, and check accessibility criteria. DeepThought is an internal data science platform for developing and deploying models for better product design, data-driven decisions, and risk management. The Nebula track focuses on managing large data volumes and delivering a scalable architecture, with a focus on Data Mesh Architecture Concept. Vanguard involves the design of domain-specific data hubs, with modeling activities completed as of December 2020. Bedrock includes a set of tools implemented as part of the Tesseract Platform, such as Ingestion, Presto, Collibra, Apache Zeppelin, Apache Ranger, and DeepThought. Tesseract also offers curated and certified data sets, Data as a Service, ML as a Service, and an Observability Dashboard."}, {"You:": "summarize this in 50 words", "LLM:": "VMware uses various internal platforms for data management and analytics. Iceberg is the main portal for data needs and research. DeepThought is used for data-driven decisions and operational efficiencies. Hubble focuses on data monitoring, while HawkEye provides tools for data storytelling. Vanguard designs domain-specific data hubs, and Nebula manages data volumes."}]

app.py
import streamlit as st
from PyPDF2 import PdfReader
from streamlit_extras.add_vertical_space import add_vertical_space
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.vectorstores import FAISS
from langchain.llms import OpenAI
from langchain.chains.question_answering import load_qa_chain
import pickle
from dotenv import load_dotenv
import os
from langchain.callbacks import get_openai_callback
import json

# side bar
with st.sidebar:
    st.title('PDF chat App')
    st.markdown("""
    # Capabilities of App
    Reads PDF and 
    Answers Question.
    """)
    add_vertical_space(5)
    st.write('Madness is like Gravity.. all you need is a push.')


def chat_hist(file_path,chat_dict_list):
    if os.path.exists(file_path):
    # check size
        file_size = os.path.getsize(file_path)
        if file_size ==0:
            print('history file is empty')
            hist = chat_dict_list
            with open(file_path, "w") as file:
                json.dump(hist, file)
        else:
            print('history already present')
            with open(file_path, "r") as file:
                loaded_list_of_dicts = json.load(file)
            new_chat = chat_dict_list
            new_hist = loaded_list_of_dicts + new_chat
            with open(file_path, "w") as file:
                json.dump(new_hist, file) 

def clear_file_content(file_path):
    with open(file_path, "w") as file:
        file.truncate(0)

def main():
    st.header("Chat with PDF")
    load_dotenv()
    file_path = "history.txt"
    
    #upload PDF
    pdf = st.file_uploader("Upload your PDF",type='pdf')
    
    #read pdf only when available
    if pdf is not None:
        pdf_reader = PdfReader(pdf)
        pdf_name = pdf.name[:-4]
        #st.write(pdf_reader)
        text = ""
        
        for page in pdf_reader.pages:
            text = text+page.extract_text()
        #st.write(text)

        text_splitter = RecursiveCharacterTextSplitter(
            chunk_size = 1000,
            chunk_overlap = 200,
            length_function=len
            )
        chunks = text_splitter.split_text(text=text)

        #st.write(chunks)
        
        if os.path.exists(f"{pdf_name}.pkl"):
            with open(f"{pdf_name}.pkl", 'rb') as f:
                vectorStore = pickle.load(f)
            st.write("Embedding already available for pdf.")
        else:
            #embeddings object
            st.write("Using OpenAi to create embeddings.")
            embeddings = OpenAIEmbeddings()
            vectorStore = FAISS.from_texts(chunks, embedding=embeddings)
            with open(f"{pdf_name}.pkl","wb") as f:
                pickle.dump(vectorStore,f)

        # question box
        st.write("Howdy Partner!..What do you want to know about that PDF you uploaded:")
        query = st.text_input("Type your question:")
        #st.write(query)

        chat_history=[]
        if query:
            print('in while')
            docs = vectorStore.similarity_search(query, k=3)

            llm = OpenAI(model_name = 'gpt-3.5-turbo',temperature=0)
            chain = load_qa_chain(llm=llm, chain_type = "stuff")
            with get_openai_callback() as cb:
                response = chain.run(input_documents = docs, question=query)
                print(cb)
            st.write(response)
            # query = st.text_input("Type your question:")
            # continue
            chat_history.append({"You:":query,"LLM:":response})
            chat_hist(file_path,chat_history)
            # print(chat_history)

            st.subheader("Chat History")
            with open(file_path,'r') as file:
                loaded_hist = json.load(file)
            for item in loaded_hist:
                st.write(f"You: {item['You:']}")
                st.write(f"LLM: {item['LLM:']}")
                       
    else:
        st.write("No PDF uploaded")
    

if __name__ =='__main__':
    # print('creating ch')
    # chat_history=[]
    print('running main')
    main()
    #print(chat_history)

app_v1.py
import streamlit as st
from PyPDF2 import PdfReader
from streamlit_extras.add_vertical_space import add_vertical_space
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.vectorstores import FAISS
from langchain.llms import OpenAI
from langchain.chains.question_answering import load_qa_chain
import pickle
from dotenv import load_dotenv
import os
from langchain.callbacks import get_openai_callback
import json

# side bar
with st.sidebar:
    st.title('PDF chat App')
    st.markdown("""
    # Capabilities of App
    Reads PDF and 
    Answers Question.
    """)
    add_vertical_space(5)
    st.write('Madness is like Gravity.. all you need is a push.')


def chat_hist(file_path,chat_dict_list):
    if os.path.exists(file_path):
    # check size
        file_size = os.path.getsize(file_path)
        if file_size ==0:
            print('history file is empty')
            hist = chat_dict_list
            with open(file_path, "w") as file:
                json.dump(hist, file)
        else:
            print('history already present')
            with open(file_path, "r") as file:
                loaded_list_of_dicts = json.load(file)
            new_chat = chat_dict_list
            new_hist = loaded_list_of_dicts + new_chat
            with open(file_path, "w") as file:
                json.dump(new_hist, file) 

def clear_file_content(file_path):
    with open(file_path, "w") as file:
        file.truncate(0)

def main():
    st.header("Chat with PDF")
    load_dotenv()
    file_path = "history.txt"
    
    #upload PDF
    pdf = st.file_uploader("Upload your PDF",type='pdf')
    
    #read pdf only when available
    if pdf is not None:
        pdf_reader = PdfReader(pdf)
        pdf_name = pdf.name[:-4]
        #st.write(pdf_reader)
        text = ""
        
        for page in pdf_reader.pages:
            text = text+page.extract_text()
        #st.write(text)

        text_splitter = RecursiveCharacterTextSplitter(
            chunk_size = 1000,
            chunk_overlap = 200,
            length_function=len
            )
        chunks = text_splitter.split_text(text=text)

        #st.write(chunks)
        
        if os.path.exists(f"{pdf_name}.pkl"):
            with open(f"{pdf_name}.pkl", 'rb') as f:
                vectorStore = pickle.load(f)
            st.write("Embedding already available for pdf.")
        else:
            #embeddings object
            st.write("Using OpenAi to create embeddings.")
            embeddings = OpenAIEmbeddings()
            vectorStore = FAISS.from_texts(chunks, embedding=embeddings)
            with open(f"{pdf_name}.pkl","wb") as f:
                pickle.dump(vectorStore,f)

        # question box
        st.write("Howdy Partner!..What do you want to know about that PDF you uploaded:")
        query = st.text_input("Type your question:")
        #st.write(query)

        chat_history=[]
        if query:
            print('in while')
            docs = vectorStore.similarity_search(query, k=3)

            llm = OpenAI(model_name = 'gpt-4',temperature=0)
            chain = load_qa_chain(llm=llm, chain_type = "stuff")
            with get_openai_callback() as cb:
                response = chain.run(input_documents = docs, question=query)
                print(cb)
            st.write(response)
            # query = st.text_input("Type your question:")
            # continue
            chat_history.append({"You:":query,"LLM:":response})
            chat_hist(file_path,chat_history)
            # print(chat_history)

            st.subheader("Chat History")
            with open(file_path,'r') as file:
                loaded_hist = json.load(file)
            for item in loaded_hist:
                st.write(f"You: {item['You:']}")
                st.write(f"LLM: {item['LLM:']}")
                       
    else:
        st.write("No PDF uploaded")
    

if __name__ =='__main__':
    # print('creating ch')
    # chat_history=[]
    print('running main')
    main()
    #print(chat_history)

app_v3.py
import streamlit as st
from PyPDF2 import PdfReader
from streamlit_extras.add_vertical_space import add_vertical_space
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.vectorstores import FAISS
from langchain.llms import OpenAI
from langchain.chains.question_answering import load_qa_chain
from langchain.document_loaders.image import UnstructuredImageLoader
import pickle
from dotenv import load_dotenv
import os
from langchain.callbacks import get_openai_callback
import json

import pypdfium2 as pdfium
import matplotlib.pyplot as plt
from PIL import Image
from io import BytesIO
from pytesseract import image_to_string

# side bar
with st.sidebar:
    st.title('PDF chat App')
    st.markdown("""
    # Capabilities of App
    Reads PDF and 
    Answers Question.
    """)
    add_vertical_space(5)
    st.write("""It's not who I am .. But what I do that defines me. --Batman""")

def convert_pdf_to_images(pdf_file, scale=300/72):

    pdf = pdfium.PdfDocument(pdf_file)
    page_indices = [i for i in range(len(pdf))]

    renderer = pdf.render(
        pdfium.PdfBitmap.to_pil,
        page_indices = page_indices,
        scale = scale
    )

    final_images =[]

    for i, image in zip(page_indices, renderer):
        image_byte_array = BytesIO()
        image.save(image_byte_array, format='jpeg', optimize=True)
        image_byte_array = image_byte_array.getvalue()
        final_images.append(dict({i:image_byte_array}))

    return final_images

def extract_text_from_image(list_dict_final_images):
    image_list = [list(data.values())[0] for data in list_dict_final_images]
    text = ""
    
    for index, image_bytes in enumerate(image_list):
        image = Image.open(BytesIO(image_bytes))
        raw_text = str(image_to_string(image))
        text = text+raw_text
    
    return text

def split_text_to_chunks(text):
    text_splitter = RecursiveCharacterTextSplitter(
            chunk_size = 1000,
            chunk_overlap = 200,
            length_function=len
            )
    chunks = text_splitter.split_text(text=text)
    return chunks
    
def create_vectors(chunks):
    pdf_name = 'All_Hands_2023'
    if os.path.exists(f"{pdf_name}.pkl"):
        with open(f"{pdf_name}.pkl", 'rb') as f:
            vectorStore = pickle.load(f)
        st.write("Embedding already available for pdf.")
        return vectorStore
    else:
        #embeddings object
        st.write("Using OpenAi to create embeddings.")
        embeddings = OpenAIEmbeddings()
        vectorStore = FAISS.from_texts(chunks, embedding=embeddings)
        with open(f"{pdf_name}.pkl","wb") as f:
            pickle.dump(vectorStore,f)
        return vectorStore

def chat_hist(file_path,chat_dict_list):
    if os.path.exists(file_path):
    # check size
        file_size = os.path.getsize(file_path)
        if file_size ==0:
            print('history file is empty')
            hist = chat_dict_list
            with open(file_path, "w") as file:
                json.dump(hist, file)
        else:
            print('history already present')
            with open(file_path, "r") as file:
                loaded_list_of_dicts = json.load(file)
            new_chat = chat_dict_list
            new_hist = loaded_list_of_dicts + new_chat
            with open(file_path, "w") as file:
                json.dump(new_hist, file) 

def main():

    st.header("Chat with PDF")
    load_dotenv()
    file_path = "history.txt"
    convert_pdf_to_img = convert_pdf_to_images('All_Hands_2023.pdf')
    final_text = extract_text_from_image(convert_pdf_to_img)
    text_chunks = split_text_to_chunks(final_text)
    vec = create_vectors(text_chunks)
        
    # question box
    st.write("Howdy Partner!..What do you want to know about that PDF you uploaded:")
    query = st.text_input("Type your question:")
    #st.write(query)

    chat_history=[]
    if query:
        print('in while')
        docs = vec.similarity_search(query, k=3)

        llm = OpenAI(model_name = 'gpt-3.5-turbo',temperature=0)
        chain = load_qa_chain(llm=llm, chain_type = "stuff")
        with get_openai_callback() as cb:
            response = chain.run(input_documents = docs, question=query)
            print(cb)
        st.write(response)
        # query = st.text_input("Type your question:")
        # continue
        chat_history.append({"You:":query,"LLM:":response})
        chat_hist(file_path,chat_history)
        # print(chat_history)

        st.subheader("Chat History")
        with open(file_path,'r') as file:
            loaded_hist = json.load(file)
        for item in loaded_hist:
            st.write(f"You: {item['You:']}")
            st.write(f"LLM: {item['LLM:']}")
        
    
if __name__ =='__main__':
    print('running main')
    main()
    

app_v2.py
import streamlit as st
import PyPDF2 as pp 
from PyPDF2 import PdfReader
from streamlit_extras.add_vertical_space import add_vertical_space
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings.openai import OpenAIEmbeddings
from langchain.vectorstores import FAISS
from langchain.llms import OpenAI
from langchain.chains.question_answering import load_qa_chain
import pickle
from dotenv import load_dotenv
import os
from langchain.callbacks import get_openai_callback
import json
import glob

# side bar
with st.sidebar:
    st.title('PDF chat App')
    st.markdown("""
    # Capabilities of App
    Reads PDF and 
    Answers Question.
    """)
    add_vertical_space(5)
    st.write('Madness is like Gravity.. all you need is a push.')


def chat_hist(file_path,chat_dict_list):
    if os.path.exists(file_path):
    # check size
        file_size = os.path.getsize(file_path)
        if file_size ==0:
            print('history file is empty')
            hist = chat_dict_list
            with open(file_path, "w") as file:
                json.dump(hist, file)
        else:
            print('history already present')
            with open(file_path, "r") as file:
                loaded_list_of_dicts = json.load(file)
            new_chat = chat_dict_list
            new_hist = loaded_list_of_dicts + new_chat
            with open(file_path, "w") as file:
                json.dump(new_hist, file) 

def pocessing_hist(processed_pdf_path, processed_pdf_list):
    if os.path.exists(processed_pdf_path):
    # check size
        file_size = os.path.getsize(processed_pdf_path)
        if file_size ==0:
            print('processing pdf history file is empty')
            hist = processed_pdf_list
            with open(processed_pdf_path, "w") as file:
                json.dump(hist, file)
        else:
            print('processing pdf history already present')
            with open(processed_pdf_path, "r") as file:
                loaded_list_of_pdfs = json.load(file)
            new_pdf_list = processed_pdf_list
            new_pdf_list = loaded_list_of_pdfs + new_pdf_list
            with open(processed_pdf_path, "w") as file:
                json.dump(new_pdf_list, file)

def clear_file_content(file_path):
    with open(file_path, "w") as file:
        file.truncate(0)

def main():
    st.header("Chat with PDF")
    load_dotenv()
    file_path = "history.txt"
    processed_pdf_path = "processed_pdf.txt"
    
    #upload PDF
    # pdf = st.file_uploader("Upload your PDF",type='pdf')
    
    pdf_dir = '/home/shashank/ml_app'
    
    pdf_pattern = os.path.join(pdf_dir,'*.pdf')

    pdf_files = glob.glob(pdf_pattern)

    pdf_processed=[]
    pdf_content =[]
    loaded_list_of_pdfs=[]

    text = ""

    # checking previously processed PDFs

    processing_file_size = os.path.getsize(processed_pdf_path)
    if processing_file_size >0:
        with open(processed_pdf_path, "r") as file:
            loaded_list_of_pdfs = json.load(file)
            loading_hist = True

    for file in pdf_files:
        #st.write("processing :",file)
        pdf_name = file[len(pdf_dir)+1:-4]
        #st.write(pdf_name)
        pdf_processed.append(pdf_name)
        pdf_path = file
        
        if pdf_name not in loaded_list_of_pdfs:
            st.write("Processing new PDFs")
            with open(pdf_path,'rb') as pdf:
                pdf_reader = pp.PdfReader(pdf)

                for page in pdf_reader.pages:
                    page_text = page.extract_text()
                    pdf_content.append(page_text)
                    # text = text+page.extract_text()
        else:
            st.write("No files to process")
    # st.write(len(pdf_content))
    text =''.join(pdf_content)
    # st.write(text)

    text_splitter = RecursiveCharacterTextSplitter(
            chunk_size = 1000,
            chunk_overlap = 200,
            length_function=len
            )
    chunks = text_splitter.split_text(text=text)

    pdf_vec_file_name = '_'.join(pdf_processed)

    if os.path.exists(f"{pdf_vec_file_name}.pkl"):
        with open(f"{pdf_vec_file_name}.pkl", 'rb') as f:
            vectorStore = pickle.load(f)
        st.write("Embedding already available for pdf.")
    else:
        st.write("Using OpenAi to create embeddings.")
        embeddings = OpenAIEmbeddings()
        vectorStore = FAISS.from_texts(chunks, embedding=embeddings)
        with open(f"{pdf_vec_file_name}.pkl","wb") as f:
            pickle.dump(vectorStore,f)

    # saving list of pdfs processed
    processing_file_size = os.path.getsize(processed_pdf_path)
    if processing_file_size >0:
        with open(processed_pdf_path, "r") as file:
            loaded_list_of_pdfs = json.load(file)
        if pdf_processed != loaded_list_of_pdfs:
            st.write("new pdfs")
            pocessing_hist(processed_pdf_path,pdf_processed)
        else: 
            st.write("No new PDFs")
    else:
        st.write("Processing history file is empty")
        pocessing_hist(processed_pdf_path,pdf_processed)

    # question box
    context = ';'.join(pdf_processed)
    st.write("Context is:")
    st.write(context)
    st.write("Howdy Partner!..What do you want to know about that PDFs mentioned above.")
    query = st.text_input("Type your question:")
    
    chat_history=[]
    if query:
        print('in while')
        docs = vectorStore.similarity_search(query, k=3)

        llm = OpenAI(model_name = 'gpt-3.5-turbo',temperature=0)
        chain = load_qa_chain(llm=llm, chain_type = "stuff")
        with get_openai_callback() as cb:
            response = chain.run(input_documents = docs, question=query)
            print(cb)
        st.write(response)
        # query = st.text_input("Type your question:")
        # continue
        chat_history.append({"You:":query,"LLM:":response})
        chat_hist(file_path,chat_history)
        # print(chat_history)

        st.subheader("Chat History")
        with open(file_path,'r') as file:
            loaded_hist = json.load(file)
        for item in loaded_hist:
            st.write(f"You: {item['You:']}")
            st.write(f"LLM: {item['LLM:']}")
   

if __name__ =='__main__':
    # print('creating ch')
    # chat_history=[]
    print('running main')
    main()
    #print(chat_history)